tests/test_checksum.c

```diff
@@ -178,6 +178,103 @@ START_TEST(test_cached_checksum)
 }
 END_TEST
 
+START_TEST(test_cached_checksum_value)
+{
+    FILE *f;
+    int fd, ret;
+    gboolean checksum_ret, matches;
+    ssize_t attr_ret;
+    char *filename;
+    static char *expected = "d78931fcf2660108eec0d6674ecb4e02401b5256a6b5ee82527766ef6d198c67";
+    struct stat st;
+    char buf[256];
+    GError *tmp_err = NULL;
+    gchar *timestamp_key = g_strconcat(XATTR_CHKSUM_PREFIX, "mtime", NULL);
+    gchar *checksum_key = g_strconcat(XATTR_CHKSUM_PREFIX, "sha256", NULL);
+    gchar *mtime_str = NULL;
+    gchar *calculated = NULL;
+
+    filename = lr_pathconcat(test_globals.tmpdir, "/test_checksum_value", NULL);
+    f = fopen(filename, "w");
+    fail_if(f == NULL);
+    fwrite("foo\nbar\n", 1, 8, f);
+    fclose(f);
+
+    // Assert no cached checksum exists
+    attr_ret = GETXATTR(filename, timestamp_key, &buf, sizeof(buf)-1);
+    fail_if(attr_ret != -1);  // Cached timestamp should not exists
+    attr_ret = GETXATTR(filename, checksum_key, &buf, sizeof(buf)-1);
+    fail_if(attr_ret != -1);  // Cached checksum should not exists
+
+    // Calculate checksum
+    fd = open(filename, O_RDONLY);
+    fail_if(fd < 0);
+    checksum_ret = lr_checksum_fd_compare(LR_CHECKSUM_SHA256,
+                                      fd,
+                                      "",
+                                      1,
+                                      &matches,
+                                      &calculated,
+                                      &tmp_err);
+    fail_if(tmp_err);
+    fail_if(!checksum_ret);
+    // We pass in an empty string for expected, so we must not match.
+    fail_if(matches);
+    close(fd);
+    fail_if(strcmp(calculated, expected));
+    g_free(calculated);
+    calculated = NULL;
+
+    // Assert cached checksum exists
+    attr_ret = GETXATTR(filename, checksum_key, &buf, sizeof(buf)-1);
+
+    if (attr_ret == -1) {
+        // Error encountered
+        if (errno == ENOTSUP) {
+            // Extended attributes are not supported
+            goto exit_label;
+        }
+        // Any other errno means fail
+        fail_if(attr_ret == -1);
```

Comment:
Testing of xattrs is tricky and two weaks ago I had tough times with it so I can show off a bit ;) . Long story short - the test will fail on this line in case the test file is created on regular filesystem - not tmpfs which /tmp/ usually is.

- xattrs are supported on tmpfs but user extended attributes are not permitted (see man tmpfs)
- so if you try `make test` on your machine the test will pass - because the test file is created inside the `/tmp/` dir which is probably a tmpf, you get ENOTSUP error on line 233 and the rest of the test is skipped.
- but if you create the test file on a regular filesystem (either by changing the location on the line 197 or by trying to build a rpm from this PR using mock - that is how I discovered the problem) you find out that the test doesn't work. Instead of `ENOTSUP` you start getting `ENODATA`.
- the reason is another bug in librepo - the computed checksum is written to the xattr only when expected and calculated checsum matched - see `librepo/checksum.c:274`. Once calculated the checksum should be written to xattr regardless it matches.

I didn't solve the issue with tests and tmpfs - it cannot be 100% solved but we probably  could minimize its impact by creating testing files somewhere inside the build directory - passing it to to `test_main` using current `run_tests.sh` the same way as the repos directory is passed? But this definitely might be done later in other PR.

Reply:
This is a very good catch. Given the current semantics, I think it's better to omit the xattr checks (because we'll never have `expected == calculated`). I'm keen to get this minor change in, because it's objectively less wrong than before. I will update this test, and omit this incorrect code for now.

Reply:
See #235 for the caching bug. Note this isn't as simple as I originally thought.
#236 for the xattr issue 

If these are dupes of existing issues, please feel free to close them.